### PR TITLE
feat: Bound Method Realm support

### PR DESCRIPTION
**Pull request by jaekwon**
When a function is called, it's usually enough to get the pkgpath from where the function is declared, to find out whether we are switching to a new realm or not.

But when the function is a bound method, and the method and declared type are defined in a /p/ package, we need to look at the owning (persisting) pkgpath of the receiver, if any. When one is found, it uses that. 

There are some edge cases to consider, I'm not sure if they are all fixed, but we will find out.
Here are some that are considered;
1. if the receiver is an unpersisted new object with method, declared in a p module, it has no realm until it becomes persisted.
2. nil receivers aren't objects, so even if it were stored in a realm, it would be stored by "reference" from a pointervalue which also is not an object, so there's no realm to consider even after persistence.
3. if the method from 1 makes any modifications to the /p/ package's block state, it should result in an error and the tx should fail.
4. if a /p/ declared thing with method gets stored in realm X, it should be possible for realm Y to call it, and have it modify realm X (e.g. say realm X attached a closure to the thing).

### CODE REVIEW

File: gnovm/pkg/gnolang/ownership.go [hunk 1]
```diff
@@ -332,6 +332,7 @@ func (tv *TypedValue) GetFirstObject(store Store) Object {
 		// something in it; in that case, ignore the base.  That will
 		// likely require maybe a preparation step in persistence
 		// ( or unlikely, a second type of ref-counting).
+		// XXX is there an issue with Base=nil pointers here cross realm?
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/ownership.go:335]: TODO

File: gnovm/pkg/gnolang/machine.go [hunk 2]
```diff
@@ -1627,8 +1631,24 @@ func (m *Machine) PushFrameCall(cx *CallExpr, fv *FuncValue, recv TypedValue) {
 	}
 	m.Package = pv
 	rlm := pv.GetRealm()
+	if rlm == nil && recv.IsDefined() {
+		// if bound method, get realm from receiver.
+		obj := recv.GetFirstObject(m.Store)
+		if obj == nil {
+			// panic("XXX not sure why this would be")
+			fmt.Println("XXX XXX", recv.String())
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/machine.go:1639]: TODO

File: gnovm/tests/files/zrealm_crossrealm13.gno [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+func main() {
+	// even though we are running within a realm,
+	// we aren't storing the result of crossrealm.Make1(),
+	// so this should print fine.
+	crossrealm.Make1().Touch().Print()
+}
+
+// Output:
+// A: 2
+// B: LocalStruct{123}
```
Comment (jaekwon) [inline gnovm/tests/files/zrealm_crossrealm13.gno:17]: TODO write some more tests.

**Comment by albttx** [comment]
Any update on this @jaekwon / @moul ?

**Comment by thehowl** [comment]
txtar created by albttx which illustrates the kind of behaviour that should be fixed by this PR (or any continuation of this work):

```
# Set up GNOROOT in the current directory.
mkdir $WORK/gnovm
symlink $WORK/gnovm/stdlibs -> $GNOROOT/gnovm/stdlibs
env GNOROOT=$WORK

gno test -verbose ./examples/gno.land/r/demo/realm2

stderr '=== RUN   TestDo'
stderr '--- PASS: TestDo.*'

stderr '=== RUN   file/realm2_filetest.gno'
stderr '--- PASS: file/realm2_filetest.*'

-- examples/gno.land/p/demo/counter/gno.mod --
module gno.land/p/demo/counter

-- examples/gno.land/p/demo/counter/counter.gno --
package counter

type Counter struct {
	n int
}

func (c *Counter) Inc() {
	c.n++
}

-- examples/gno.land/r/demo/realm1/realm1.gno --
package realm1

import "gno.land/p/demo/counter"

var c = counter.Counter{}

func GetCounter() *counter.Counter {
	return &c
}

-- examples/gno.land/r/demo/realm2/realm2.gno --
package realm2

import (
	"gno.land/r/demo/realm1"
)

func Do() {
	realm1.GetCounter().Inc()
}

-- examples/gno.land/r/demo/realm2/realm2_filetest.gno --
// PKGPATH: gno.land/r/tests
package tests

import "gno.land/r/demo/realm2"

func main() {
	realm2.Do()
	println("OK")
}

// Output:
// OK

-- examples/gno.land/r/demo/realm2/realm2_test.gno --
package realm2

import "testing"

func TestDo(t *testing.T) {
	Do()
}
```

**Review by moul — APPROVED** [review]

**Review by moul — APPROVED** [review]

**Pull request merged**